readlog/readlog.1
.TH READLOG 1 "19 November 2004"
.\" (C) Copyright 2002-2004 Diomidis Spinellis.  All rights reserved.
.\"
.\" Permission to use, copy, and distribute this software and its
.\" documentation for any purpose and without fee for noncommercial use
.\" is hereby granted, provided that the above copyright notice appear in
.\" all copies and that both that copyright notice and this permission notice
.\" appear in supporting documentation.
.\"
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
.\" WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
.\" MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" $Id: readlog.1,v 1.3 2004-11-19 14:54:28 dds Exp $
.\"
.SH NAME
readlog \- text-based access to the Windows event log
.SH SYNOPSIS
\fBreadlog\fP
[\fB\-t\fP \fIfmt\fP]
[\fB\-v\fP \fIsrv\fP]
[\fB\-riuwsycabdn\fP]
[\fIsource ...\fP]
.SH DESCRIPTION
\fIReadlog\fP
provides text-based access to the Windows event log.
It can thus be used to textually process the data that is normaly
seen through the Windows \fIevent viewer\fP program.
Running \fIreadlog\fP without any options will generate a listing
of the \fISystem\fP event log in a format remininscent of the Unix
\fIsyslogd\fP log files such as the following:
.br
Apr 20 08:35:28 SEAGULL Srv: -: Warning: The C: disk is at or near capacity.  You may need to delete some files.
.br
By default
each entry contains the time, the computer name, the application name,
the log message category (where available), the log message type
(error, warning, information, audit success, or audit failure),
and the the error message.
Under Windows the event log does not contain the actual messages, but
pointers to files that contain pre-compiled message strings.
Failure to obtain such a message string will result in an error message,
but \fIreadlog\fP will continue its operation displaying the message code.
.LP
Without any parameters \fIreadlog\fP will print the contents of the
\fISystem\fP event log.
Windows systems typically also contain \fIApplication\fP and \fISecurity\fP
logs.
In addition, applications can install other custom log files.
You can see the files available on your system in the
registry under the
HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Services\\Eventlog
branch.
One or more event log files can be specified as arguments to
\fIreadlog\fP.
.SH OPTIONS
.IP "\fB\-f\fP \fBfmt\fP"
Specify the format to display the event generation time using the
\fIstrftime(3)\fP escape sequences.
.IP "\fB\-v\fP \fBsrc\fP"
Specify the server name from which to obtain the event log as a
UNC name.
.IP "\fB\-r\fP"
Print entries in reverse chronological order starting from the latest
entry and going back in time.
.IP "\fB\-i\fP"
Output the decimal event id.
.IP "\fB\-u\fP"
Do not print user information;
normally user information is printed using the \fIdomain\\user\fP convention.
.IP "\fB\-w\fP"
Do not print the workstation name.
.IP "\fB\-s\fP"
Do not print the event source.
.IP "\fB\-y\fP"
Do not print the event type.
.IP "\fB\-c\fP"
Do not print the event category (most events have no categories registered, so
you will in many cases just see a single dash).
.IP "\fB\-a\fP"
Output event-specific data as ASCII.
.IP "\fB\-b\fP"
Output event-specific data as hex bytes.
.IP "\fB\-d\fP"
Output event-specific data as hex doublewords.
.IP "\fB\-n\fP"
Format event using newline separators;
normally each event is displayed in a single line.

.SH EXAMPLE
readlog | grep "The Event log service was started" | wc -l
.br
can be used to count the number of system boots registered in the log.
.LP
readlog Application |
awk -F: "/Outbound: Information: Fax Sent/{print $12}" |
sort |
uniq -c |
sort -rn
.br
Create an list of fax recipients ordered by the number of faxes they have
received.
.SH "SEE ALSO"
D. Spinellis.  Outwit: Unix tool-based programming meets the Windows world.
In \fIUSENIX 2000 Technical Conference Proceedings\fP, pages 149-158, San Diego, CA, USA,
June 2000, USENIX Association.
.br
Microsoft Corporation.
\fIMicrosoft Windows NT Server 4.0 Resource Kit\fP.
Microsoft Press.

.SH AUTHOR
(C) Copyright 2002 Diomidis Spinellis.  All rights reserved.
.LP
Permission to use, copy, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation.
.LP
THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.SH BUGS
Remote system access has not been tested.
.LP
Windows event log messages are sometimes difficult to parse using text-based
tools.


ChangeLog.txt
Version 1.25
(March 20 2016)

winclip supports the specification of copied data as RTF

winreg supports the REG_DWORD_BIG_ENDIAN data type

odbc now supports a -n option to specify the printed value of NULL fields.
(suggested by John McMahon)

winclip now supports a -g option for copying graphics files into the
clipboard.

The connection example in the odbc application is now correct
(suggested by Oliver Peters).

winclip now supports a -i option for reporting the available data formats.

Correct memory allocation and locking in winclip. (bug fix)

-----------------------------
Version 1.24
(February 10 2005)
New options in winclip allow the specification of the language
and sublanguage of the clipboard text.
-----------------------------
Version 1.23
(November 2004)
Made readlog more robust in the face of invalid data in the registry:
- Negative event time values would cause the program to crash.
- Error reporting on missing message sources was misleading.
-----------------------------
Version 1.22
(June 2004).
Change distribution format to zip files; distributed files reside an
a correctly-named directory.
-----------------------------
Version 21
(February 2004).
Update the winclip program to support
input and output of multi-byte (UTF-8) Unicode characters and
a byte order mark.

-----------------------------
Version 20
(December 2003)
Update the winreg program to support the
names with embedded backslashes.
Problem reported by Antti Ala-heikkil&auml;.

-----------------------------
Version 19
(December 2003)
Update the winreg program to support the
QWORD registry data type and add an option to ignore errors
and continue processing.

-----------------------------
Version 18
(December 2003)
Changes by Pascal Hirsch in the odbc program
corrected error reporting and allow direct connection to ODBC drivers
without requiring the registration of an ODBC data source.
The syntax for connecting to ODBC data sources has been slightly changed.

-----------------------------
Version 17
(November 2003)
A fix by Pascal Hirsch in the odbc program
fixes two access violations, and introduces better error reporting.

-----------------------------
Version 16
(September 2003)
Corrected erroneous pasting of non-Unicode text introduced
in version 15.

-----------------------------
Version 15
(September 2003)
Corrected problem of random characters appearing at the
end of winclip output.
Noted by Robert Bassett.

-----------------------------
Version 14
(June 2003)
Winreg can connect to remote machines over the network.
Thanks to Antti Ala-Heikkil&auml;.
-----------------------------
Version 13
(Jul 12 2002)
Added the readlog command;
winreg handles additional registry entry types.
-----------------------------
Version 12
(Jan 4 2002)
Winclip supports Unicode text;
gracefully handles non-ASCII filenames.
-----------------------------
Version 11
(Sep 11 2001)
Winclip accepts filename argument.
-----------------------------
Version 10
(Jun 1 2000)
First public release.
-----------------------------
Version 09
Review release.


readlog/readlog.c
/*
 *
 * readlog - Windows log text-based access
 *
 * (C) Copyright 2002 Diomidis Spinellis
 * 
 * Permission to use, copy, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright notice appear in all copies and that
 * both that copyright notice and this permission notice appear in
 * supporting documentation.
 * 
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
 * MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 * $Id: readlog.c,v 1.4 2002-07-10 10:38:37 dds Exp $
 *
 */

#include <time.h>
#include <windows.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

/*
 * Program options
 */
static char *time_fmt = "%b %d %H:%M:%S";
static char *server = NULL;
static char *source = "System";
static int o_direction = EVENTLOG_FORWARDS_READ;
static int o_user = 1;
static int o_computer = 1;
static int o_source = 1;
static int o_type = 1;
static int o_category = 1;
static int o_ascii = 0;
static int o_hex = 0;
static int o_newline = 0;

static void
usage(char *fname)
{
	fprintf(stderr, 
		"readlog - Windows event log text-based access.  $Revision: 1.4 $\n"
		"(C) Copyright 2002 Diomidis D. Spinelllis.  All rights reserved.\n\n"

		"Permission to use, copy, and distribute this software and its\n"
		"documentation for any purpose and without fee is hereby granted,\n"
		"provided that the above copyright notice appear in all copies and that\n"
		"both that copyright notice and this permission notice appear in\n"
		"supporting documentation.\n\n"

		"THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED\n"
		"WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF\n"
		"MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.\n\n"

		"usage: %s [-t fmt] [-v server] [-ruwsycahn] [source]\n"
		"\t-t fmt\tTime format string (strftime)\n"
		"\t-v server\tServer name\n"
		"\t-r\tPrint entries in reverse order\n"
		"\t-u\tDo not print user information\n"
		"\t-w\tDo not print workstation name\n"
		"\t-s\tDo not print event source\n"
		"\t-y\tDo not print event type\n"
		"\t-c\tDo not print event category\n"
		"\t-a\tOutput event-specific data as ASCII\n"
		"\t-h\tOutput event-specific data in hex format\n"
		"\t-n\tFormat event using newline separators\n"
		"\t\tDefault source is System,\n"
		"\t\tsource can be System, Application, Security or a custom name\n"
		, fname
	);
	exit(1);
}

extern char	*optarg;	/* Global argument pointer. */
extern int	optind;		/* Global argv index. */
int getopt(int argc, char *argv[], char *optstring);
static void print_log(void);

main(int argc, char *argv[])
{
	int c;

	while ((c = getopt(argc, argv, "t:v:rucsyahn")) != EOF)
		switch (c) {
		case 't':
			if (!optarg || optarg[1])
				usage(argv[0]);
			time_fmt = optarg;
			break;
		case 'v':
			if (!optarg || optarg[1])
				usage(argv[0]);
			server = optarg;
			break;
		case 'r': o_direction = EVENTLOG_BACKWARDS_READ; break;
		case 'u': o_user = 0; break;
		case 'w': o_computer = 0; break;
		case 'c': o_category = 0; break;
		case 's': o_source = 0; break;
		case 'y': o_type = 0; break;
		case 'a': o_ascii = 1; break;
		case 'h': o_hex = 1; break;
		case 'n': o_newline = 1; break;
		case '?':
			usage(argv[0]);
		}


	if (argv[optind] != NULL) {
		if (argv[optind + 1] != NULL)
			usage(argv[0]);
		/* Dump a key */
		source = argv[optind];
	}
	print_log();
	return (0);
}



static char *
evtype(int type)
{
	switch (type) {
	case EVENTLOG_ERROR_TYPE: return "Error";
	case EVENTLOG_WARNING_TYPE: return "Warning";
	case EVENTLOG_INFORMATION_TYPE: return "Information";
	case EVENTLOG_AUDIT_SUCCESS: return "Success";
	case EVENTLOG_AUDIT_FAILURE: return "Failure";
	}
}

static void
wperror(char *s, LONG err)
{
	LPVOID lpMsgBuf;

	FormatMessage( 
	    FORMAT_MESSAGE_ALLOCATE_BUFFER | 
	    FORMAT_MESSAGE_FROM_SYSTEM | 
	    FORMAT_MESSAGE_IGNORE_INSERTS,
	    NULL,
	    err,		// GetLastError() does not seem to work reliably
	    MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), // Default language
	    (LPTSTR) &lpMsgBuf,
	    0,
	    NULL 
	);
	fprintf(stderr, "%s: %s\n", s, lpMsgBuf);
	LocalFree(lpMsgBuf);
}

static void
print_time(EVENTLOGRECORD *pelr)
{
	char buff[512];

	if (!*time_fmt)
		return;
	strftime(buff, sizeof(buff), time_fmt, localtime((time_t *)(&(pelr->TimeGenerated))));
	printf("%s ", buff);
}

void
print_uname(EVENTLOGRECORD *pelr)
{
	PSID            lpSid;
	char            szName[256];
	char            szDomain[256];
	SID_NAME_USE    snu;
	DWORD           dwLen;
	DWORD           cbName = 256;
	DWORD           cbDomain = 256;

	if (!o_user || pelr->UserSidLength == 0)
		return;
	// Point to the SID. 
	lpSid = (PSID) ((LPBYTE) pelr + pelr->UserSidOffset);
	if (LookupAccountSid(NULL, lpSid, szName, &cbName, szDomain, &cbDomain, &snu))
		printf("%s\\%s: ", szDomain, szName);
	else
		wperror("LookupAccountSid", GetLastError());
}

static char *
get_message(char *msgfile, DWORD id, char *argv[])
{
	HINSTANCE mh;
	char *outmsg;

	/* TODO: cache opened libraries */
	if ((mh = LoadLibraryEx(msgfile, NULL, LOAD_LIBRARY_AS_DATAFILE)) == NULL) {
		wperror(msgfile, GetLastError());
		return NULL;
	}
	if (FormatMessage(
		FORMAT_MESSAGE_ALLOCATE_BUFFER | 
		FORMAT_MESSAGE_FROM_HMODULE |
		FORMAT_MESSAGE_ARGUMENT_ARRAY,
		mh, id, 0, (LPTSTR)&outmsg, 1024, argv) == 0) {
		wperror("FormatMessage", GetLastError());
		FreeLibrary(mh);
		return NULL;
	}
	FreeLibrary(mh);
	return (outmsg);
}
static int
get_regentry(HANDLE rh, char *name, char *result)
{
	DWORD type;
	DWORD datalen;
	char tmpstr[1024];
	int ret;

	datalen = sizeof(tmpstr);
	if ((ret = RegQueryValueEx(rh, name, NULL, &type, tmpstr, &datalen)) != ERROR_SUCCESS)
		return 0;
	if (ExpandEnvironmentStrings(tmpstr, result, sizeof(tmpstr)) == 0)
		return 0;
	return 1;
}

static void
print_oneline(char *outmsg)
{
	char *p;

	for (p = outmsg; *p; p++)
		switch (*p) {
		case '\r': 
			break;
		case '\n': 
			if (p[1])
				putchar(' ');
			break;
		default: putchar(*p);
		}
}

static void
print_msg(EVENTLOGRECORD *pelr)
{
	char key[1024];
	char msgfile[1024];
	char paramfile[1024];
	int ret;
	HANDLE rh;
	char *outmsg;
	char *argv[1024];
	int i;
	char *p;

	print_time(pelr);
	/* Computer name */
	printf("%s ", (char *)pelr + strlen((char *)pelr + sizeof(EVENTLOGRECORD)) + sizeof(EVENTLOGRECORD) + 1);
	/* Access the registry message files */
	sprintf(key, "SYSTEM\\CurrentControlSet\\Services\\Eventlog\\%s\\%s", source, (LPSTR)((LPBYTE) pelr + sizeof(EVENTLOGRECORD)));
	if ((ret = RegOpenKeyEx(HKEY_LOCAL_MACHINE, key, 0, KEY_READ, &rh)) != ERROR_SUCCESS) {
		wperror(key, ret);
		return;
	}
	if (!get_regentry(rh, "EventMessageFile", msgfile)) {
		wperror("EventMessageFile", ret);
		RegCloseKey(rh);
		return;
	}
	/* Event source */
	if (o_source)
		printf("%s: ", (LPSTR) ((LPBYTE) pelr + sizeof(EVENTLOGRECORD)));
	/* Category */
	if (o_category) {
		char *category;
		char catfile[1024];

		if (pelr->EventCategory == 0)
			printf("-: ");
		else if (get_regentry(rh, "CategoryMessageFile", catfile) &&
			 (category = get_message(catfile, pelr->EventCategory, NULL))) {
			print_oneline(category);
			printf(": ");
			LocalFree(category);
		} else
			wperror("Event category", GetLastError());
	}
	/* Event type */
	if (o_type)
		printf("%s: ", evtype(pelr->EventType));
	RegCloseKey(rh);
	print_uname(pelr);
	/*
	 * Some message file specs use a ; separated path.
	 * This is undocumented and we do not support it.
	 */
	for (i = 0, p = (char *)((LPBYTE) pelr + pelr->StringOffset); i < pelr->NumStrings; p += strlen(argv[i]) + 1, i++)
		argv[i] = p;
	if ((outmsg = get_message(msgfile, pelr->EventID, argv)) == NULL)
		return;
	if (o_newline)
		printf("\n%s", outmsg);
	else
		print_oneline(outmsg);
	putchar('\n');
	LocalFree(outmsg);
}

static void
print_log(void)
{
	HANDLE          h;
	EVENTLOGRECORD *pevlr;
	BYTE            *bBuffer;
	DWORD           dwRead,
	                dwNeeded,
	                cRecords,
			dwSize,
	                dwThisRecord = 0;

	h = OpenEventLog(server, source);
	if (h == NULL) {
		wperror("Unable to open event log", GetLastError());
		exit(1);
	}

	bBuffer = malloc(dwSize = 8192);
again:
	pevlr = (EVENTLOGRECORD *)bBuffer;
	while (ReadEventLog(h,	// event log handle 
			    o_direction |	// reads forward 
			    EVENTLOG_SEQUENTIAL_READ,	// sequential read 
			    0,	// ignored for sequential reads 
			    pevlr,	// pointer to buffer 
			    dwSize,	// size of buffer 
			    &dwRead,	// number of bytes read 
			    &dwNeeded))	// bytes in next record 
	{
		while (dwRead > 0) {
			print_msg(pevlr);
			dwRead -= pevlr->Length;
			pevlr =
			    (EVENTLOGRECORD *) ((LPBYTE) pevlr +
						pevlr->Length);
		}
		pevlr = (EVENTLOGRECORD *)bBuffer;
	}
	switch (GetLastError()) {
	case ERROR_INSUFFICIENT_BUFFER:
		bBuffer = realloc(bBuffer, dwSize = dwNeeded);
		if (bBuffer == NULL) {
			fprintf(stderr, "Out of memory\n");
			exit(1);
		}
		goto again;
	case ERROR_HANDLE_EOF:
		break;
	default:
		wperror("ReadEventLog", GetLastError());
	}
	CloseEventLog(h);
}
